### PR TITLE
Biguint division

### LINKED ISSUE #129: Implement Big UInt Division (with remainder) for `modexp`
**Context:** `modexp.yul`

## Description: 

```solidity
/// @notice Performs the big unsigned integer square of big unsigned integers with an arbitrary amount of limbs.
/// @dev The quotient is stored from `quotientPtr` to `quotientPtr + (WORD_SIZE * nLimbs)`.
/// @dev The reminder is stored from `reminderPtr` to `reminderPtr + (WORD_SIZE * nLimbs)`.
bigUIntDivRem(dividendPtr, divisorPtr, nLimbs, quotientPtr, reminderPtr)
```

`dividendPtr`: The pointer to the MSB of the dividend.
`divisorPtr`: The pointer to the MSB of the divisor.
`nLimbs`: The number of limbs needed to represent the number to square.
`quotientPtr`: The pointer to where you want the quotient to be stored.
`remainderPtr`: The pointer to where you want the remainder to be stored.

## Subtasks:

- [x] [`sbb`](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L795) #133 
- [x] [`ct_select`](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L771) #134 
- [x] [`bitor`](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L346) #132 
- [x] [`shl`](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L512) #130 
- [x] [`shr`](https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L539) #131 

## Documentation

- [Lambdaworks `div_rem` implementation for unsigned integers](https://github.com/lambdaclass/lambdaworks/blob/8e0438adb11cdbcb404116a63f51970369102c62/math/src/unsigned_integer/element.rs#L822)
- [`crypto-bigint` implementation of div_rem](https://github.com/RustCrypto/crypto-bigint/blob/1d41918af8b585ea80978c1ae5c6e97e6a6a186d/src/uint/div.rs#L54)
- [`num-bigint` implementation of `div_rem`](https://docs.rs/num-bigint/latest/src/num_bigint/biguint/division.rs.html#119)

### PR DESCRIPTION
**Pull request by jpcenteno**
## How to test

1. Copy the function and its dependencies into a Playground.yul file.
2. Copy the following fallback in the file

```yul
// 1 limb

let nLimbs := 1
mstore(0x00, 500721)
mstore(0x20, 5)

bigUIntDivRem(0x00, 0x20, 0x700, 0x900, nLimbs, 0x40, 0x60)
console_log(mload(0x40)) // 18730
console_log(mload(0x60)) // 1

// 2 limbs
let nLimbs := 2
mstore(0x00, 0x7463719198237981235)
mstore(0x20, 0x7132857231590813759832657138957167247234980237328573573650819325)

mstore(0x40, 0x3A31B8C8CC11BCC091A)
mstore(0x60, 0xB89942B918AC8409BACC1932B89C4AB8B392391A4C011B9942B9AB9B2840C992)

bigUIntDivRem(0x00, 0x40, 0x700, 0x900, nLimbs, 0x80, 0xc0)
console_log(mload(0x80)) // 0
console_log(mload(0xa0)) // 2
console_log(mload(0xc0)) // 0
console_log(mload(0xe0)) // 1

// 3 limbs
let nLimbs := 3
mstore(0x00, 0x5bb295e9f8dbfb0a66839c44c0c0ad67e14bfd036)
mstore(0x20, 0x1d9c29443b5c29a9313f8ba267fd8296ccc4e3590b3048a0a32d19456339e8cc)
mstore(0x40, 0x45d9ceb10e1b5f51f417b314fe197fe037ba8cb143b5ff987d6652e6d1808ee4)

mstore(0x60, 0x2dd94af4fc6dfd853341ce22606056b3f0a5fe81b)
mstore(0x80, 0x0ece14a21dae14d4989fc5d133fec14b666271ac8598245051968ca2b19cf466)
mstore(0xa0, 0x22ece758870dafa8fa0bd98a7f0cbff01bdd4658a1daffcc3eb3297368c04772)

bigUIntDivRem(0x00, 0x60, 0x700, 0x900, nLimbs, 0xc0, 0x120)
console_log(mload(0xc0)) //  0
console_log(mload(0xe0)) // 0
console_log(mload(0x100)) // 2
console_log(mload(0x120)) // 0
console_log(mload(0x140)) // 0
console_log(mload(0x160)) // 0

```

There are some tests of the new aux functions created for the division operation in a comment in #129 

closes #129

### CODE REVIEW

File: precompiles/Modexp.yul [hunk 1]
```diff
@@ -11,9 +11,42 @@ object "ModExp" {
...
+            
+            /// @notice Stores a zero in big unsigned integer form in memory.
+            /// @param nLimbs The number of limbs needed to represent the operand.
+            /// @param toAddress The pointer to the MSB of the destination.
+            function zeroWithLimbSizeAt(nLimbs, toAddress) {
+                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) } {
+                    let offset := mul(i, 32)
+                    mstore(add(toAddress, offset), 0)
+                }
+            }
```
Comment (ilitteri) [inline precompiles/Modexp.yul:36]: Replacing `mul`s with `add`s is cheaper.
```suggestion
            function zeroWithLimbSizeAt(nLimbs, toAddress) {
                let offset := 0
                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) } {
                    mstore(add(toAddress, offset), 0)
                    offset := add(offset, LIMB_SIZE_IN_BYTES())
                }
            }
```

File: precompiles/Modexp.yul [hunk 2]
```diff
@@ -11,9 +11,42 @@ object "ModExp" {
...
+            /// @param nLimbs The number of limbs needed to represent the operand.
+            /// @param fromAddress The pointer to the MSB of the number to copy.
+            /// @param toAddress The pointer to the MSB of the destination.
+            function copyBigUint(nLimbs, fromAddress, toAddress) {
+                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) } {
+                    let fromOffset := add(mul(i, 32), fromAddress)
+                    let toOffset := add(mul(i, 32), toAddress)
+                    mstore(toOffset, mload(fromOffset))
+                }
+            }
```
Comment (ilitteri) [inline precompiles/Modexp.yul:48]:
```suggestion
            function copyBigUint(nLimbs, fromAddress, toAddress) {
                let offset := 0
                for { let i := 0 } lt(i, nLimbs) { i := add(i, 1) } {
                    let fromOffset := add(offset, fromAddress)
                    let toOffset := add(offset, toAddress)
                    mstore(toOffset, mload(fromOffset))
                    offset := add(offset, LIMB_SIZE_IN_BYTES())
                }
            }
```

File: precompiles/Modexp.yul [hunk 3]
```diff
@@ -402,6 +451,154 @@ object "ModExp" {
...
+                }
+
+                // At this point, `limb == limbs[i - 1]`. Where `i` equals the
+                // last value it took.
+
+                // At this point, `bitSize` equals the amount of bits in the
+                // limbs following the most significant limb.
+
+                bitSize := add(bitSize, UIntBitSize(limb))
+            }
```
Comment (ilitteri) [inline precompiles/Modexp.yul:488]:
```suggestion
            function bigUIntBitSize(basePtr, nLimbs) -> bitSize {
                bitSize := shl(8, nLimbs)

                // Iterate until finding the most significant limb or reach the end of the limbs.
                let limb := 0
                let offset := 0
                for { let i := 0 } and(lt(i, nLimbs), iszero(limb)) { i := add(i, 1) } {
                    bitSize := sub(bitSize, 256) // Decrement one limb worth of bits.
                    let ptr_i := add(basePtr, offset) // = basePtr + i * 32 bytes
                    limb := mload(ptr_i)
                    offset := add(offset, LIMB_SIZE_IN_BYTES())
                }

                // At this point, `limb == limbs[i - 1]`. Where `i` equals the
                // last value it took.

                // At this point, `bitSize` equals the amount of bits in the
                // limbs following the most significant limb.

                bitSize := add(bitSize, UIntBitSize(limb))
            }
```

File: precompiles/Modexp.yul [hunk 4]
```diff
@@ -402,6 +451,154 @@ object "ModExp" {
...
+
+            /// @notice Performs in-place `x | 1` operation.
+            /// @dev This function will mutate the memory space `mem[basePtr...(basePtr + nLimbs * 32)]`
+            /// @dev It consumes constant time, aka `O(1)`.
+            /// @param basePtr Base pointer for a big unsigned integer.
+            /// @param nLimbs Number of 32 Byte limbs composing the big unsigned integer.
+            function bigUIntInPlaceOrWith1(basePtr, nLimbs) {
+                let offset := mul(sub(nLimbs, 1), 32)
```
Comment (ilitteri) [inline precompiles/Modexp.yul:496]:
```suggestion
                let offset := shl(5, sub(nLimbs, 1))
```

**Review by ilitteri — CHANGES_REQUESTED** [review]
Looks good, left some optimization suggestions.

File: precompiles/Modexp.yul (hunk 1 above)
Comment (jpcenteno) [inline precompiles/Modexp.yul:36]:
```suggestion
            function zeroWithLimbSizeAt(nLimbs, toAddress) {
                let overflow := add(toAddress, shl(5, nLimbs))
                for { } lt(toAddress, overflow) { toAddress := add(toAddress, LIMB_SIZE_IN_BYTES()) } {
                    mstore(toAddress, 0)
                }
            }
```

Comment (ilitteri) [reply precompiles/Modexp.yul:36]: I like it, let's use this if the tests run successfully.

**Review by ilitteri — APPROVED** [review]

**Pull request merged**